[RISCV] Vectorize phi for loop carried @llvm.vp.reduce.*

NexMing opened the pull request:
LLVM vector predication reduction intrinsics return a scalar result, but on RISC-V vector reduction instructions write the result in the first element of a vector register. So when a reduction in a loop uses a scalar phi, we end up with unnecessary scalar moves:
```asm
loop:
    vmv.s.x v8, zero
    vredsum.vs v8, v10, v8
    vmv.x.s a0, v8
````
This mainly affects vector predication reduction.  This tries to vectorize any scalar phis that feed into a vector predication reduction in RISCVCodeGenPrepare, converting:
```llvm
vector.body:
%red.phi = phi i32 [ ..., %entry ], [ %red, %vector.body ]
%red = tail call i32 @llvm.vp.reduce.add.nxv4i32(i32 %red.phi, <vscale x 4 x i32> %wide.load, <vscale x 4 x i1> splat (i1 true), i32 %evl)
```
to
```llvm
vector.body:
%red.phi = phi <vscale x 2 x i32> [ ..., %entry ], [ %acc.vec, %vector.body]
%phi.scalar = extractelement <vscale x 2 x i32> %red.phi, i64 0
%acc = tail call i32 @llvm.vp.reduce.add.nxv4i32(i32 %phi.scalar, <vscale x 4 x i32> %wide.load, <vscale x 4 x i1> splat (i1 true), i32 %evl)
%acc.vec = insertelement <vscale x 2 x i32> poison, float %acc, i64 0
```
Which eliminates the scalar -> vector -> scalar crossing during instruction selection.

topperc commented:
The original patch only supported ordered fadd. Why does this to support more than just vp.fadd?

topperc commented:
> The original patch only supported ordered fadd. Why does this to support more than just vp.reduce.fadd?

Oh I guess its because of the scalar input that the regular reduce intrinsics don't have?

lukel97 commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -137,7 +137,8 @@ bool RISCVCodeGenPrepare::visitIntrinsicInst(IntrinsicInst &I) {
 
-  if (I.getIntrinsicID() != Intrinsic::vector_reduce_fadd)
+  if (I.getIntrinsicID() != Intrinsic::vector_reduce_fadd &&
+      !isa<VPReductionIntrinsic>(&I))
```
Should this be limited to just fadd reductions?

```suggestion
      !I.getIntrinsicID() != Intrinsic::vp_reduce_fadd)
```

lukel97 replied:
Oh sorry i didn't see Craig's earlier comments, that makes sense. Can you edit the doc comment for this method to mention the VP case?

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/riscv-codegenprepare-asm.ll:
```diff
@@ -42,3 +42,459 @@ vector.body:
   ret float %acc
 }
+
+define i32 @vp_reduce_add(ptr %a) {
```
Can you add these tests in a separate commit in this PR so we can see the diff before and after?

lukel97 replied:
There still only seems to be one commit in this PR

NexMing replied:
done

lukel97 replied:
Oh sorry, I meant as in only having two commits, one that adds the tests without the changes in RISCVCodgenPrepare, and another with the changes in RISCVCodegenPrepare that shows the change in the tests, e.g. see https://llvm.org/docs/InstCombineContributorGuide.html#precommit-tests

NexMing replied:
> Oh sorry, I meant as in only having two commits, one that adds the tests without the changes in RISCVCodgenPrepare, and another with the changes in RISCVCodegenPrepare that shows the change in the tests, e.g. see https://llvm.org/docs/InstCombineContributorGuide.html#precommit-tests

Sorry, I misunderstood. I get it now.

Commit:
add testcases

Commit:
[RISCV] Vectorize phi for loop carried @llvm.vp.reduce.*
        This patch is vector predication version of commit 15b0fabb21af8395c1b810e7d992a869b9ef31d8

lukel97 reviewed (approved):
LGTM, thanks

Can you update the comment above visitIntrinsicInst to say something like

```
// This affects ordered fadd reductions and VP reductions that have a scalar start value. // This tries to vectorize any scalar phis that feed into these reductions:
```

Commit:
Update the comment

NexMing commented:
Could someone help merge this PR?

lukel97 commented:
> Could someone help merge this PR?

I'm just noticing it looks like you're a member of the LLVM organization, does that mean you already have commit access?

NexMing commented:
> > Could someone help merge this PR?
> 
> I'm just noticing it looks like you're a member of the LLVM organization, does that mean you already have commit access?

I applied for permissions back when LLVM was using Phabricator, but they may have been revoked due to long-term inactivity. Do I need to reapply for the permissions now?

lukel97 commented:
> > > Could someone help merge this PR?
> > 
> > 
> > I'm just noticing it looks like you're a member of the LLVM organization, does that mean you already have commit access?
> 
> I applied for permissions back when LLVM was using Phabricator, but they may have been revoked due to long-term inactivity. Do I need to reapply for the permissions now?

Oh yes I see you've been moved to the triagers team. I think according to the [RFC](https://discourse.llvm.org/t/rfc2-new-criteria-for-commit-access/77110#regaining-commit-access-3) you can create a new GitHub issue to restore commit access? 

@tstellar I couldn't find anything in https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access about regaining commit access as described in the RFC, should that be documented somewhere?
